[FIX] 멘토링 카드 조회 응답에 리뷰 평점과 리뷰 개수 추가 구현

dye0p opened the pull request:
## Issue Number
closed #477

## As-Is
<!-- 문제 상황 정의 -->
- 멘토링 목록 조회 시, 각 멘토링에 대한 리뷰 평점과 리뷰 개수가 없어 사용자에게 충분한 정보를 제공하지 못했습니다.


## To-Be
<!-- 변경 사항 -->
- 멘토링 목록 조회 응답(MentoringSummaryResponse)에 각 멘토링의 평균 리뷰 평점(reviewAverage)과 총 리뷰 개수(reviewCount)를 추가했습니다.

### 주요 변경 내용
**1. `MentoringRepository`에 `findReviewStats()` 추가**
- JPQL을 사용하여 각 멘토링 ID별로 평균 평점과 리뷰 수를 계산하는 쿼리를 추가했습니다.
- N+1 문제를 방지하고 데이터베이스 조회를 최적화하기 위해, 모든 멘토링의 리뷰 통계를 한 번의 쿼리로 가져옵니다.

**2. `ReviewStats` DTO 생성**
- mentoringId, reviewAverage, reviewCount를 담는 새로운 Record DTO를 생성하여 쿼리 결과를 효율적으로 관리합니다.

**3. `MentoringService` 로직 수정**
- findMentoringSummaries 메서드에서 findReviewStats()를 호출하여 리뷰 통계 정보를 가져옵니다.
- 가져온 통계 정보를 `Map<Long, ReviewStats>` 형태로 변환하여 각 멘토링에 대한 리뷰 정보를 쉽게 찾을 수 있도록 개선했습니다.
- 멘토링에 리뷰가 없는 경우, 평균 평점 0.0과 리뷰 개수 0을 기본값으로 사용하도록 처리했습니다.

**4. `MentoringSummaryResponse` 필드 추가**
- reviewAverage와 reviewCount 필드를 추가하여 API 응답에 리뷰 통계가 포함되도록 했습니다.

**5. 테스트 코드 수정**
- 변경된 응답 형식에 맞춰 MentoringControllerTest와 MentoringServiceTest의 테스트 케이스를 수정하고, 리뷰 데이터 생성 로직을 추가하여 기능의 정확성을 검증했습니다.

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?


## (Optional) Additional Description


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- 신기능
  - 멘토링 목록/요약에 리뷰 평균 평점과 리뷰 수가 함께 제공됩니다.
  - 리뷰가 없는 멘토링도 기본값(평점 0.0, 리뷰 수 0)으로 표시됩니다.

- 테스트
  - 리뷰/예약 데이터를 포함한 통합·서비스 테스트를 보강하여 리뷰 통계 노출을 검증했습니다.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dye0p committed:
feat: 리뷰 정보를 저장할 dto 구현

dye0p committed:
feat: 멘토링 카드정보 응답에 리뷰 정보(리뷰 평점, 리뷰 개수) 추가

dye0p committed:
refactor: 리뷰 정보 dto 네이밍 명확하게 변경(ReviewStatus -> ReviewStats)

- 통계라는 이름의 stats가 적절하다 판단

dye0p committed:
feat: 리뷰 통계정보를 반환하는 JPQL 쿼리 메서드 추가

- Review 행(row)를 가져오기 위해서 mentoring과 reservation을 모두 조회해서 가져와야 했기 때문에, 상당히 많은 양의 쿼리 발생과, 예측할 수 없는 N+1 문제가 우려 되었기에 JPQL을 사용해여 직접 쿼리를 작성하는 방식을 선택함.

dye0p committed:
feat: 멘토링 카드 조회 Service 로직에 리뷰 통계 정보도 같이 가져오도록 추가

dye0p committed:
test: 멘토링 카드 조회 기존 테스트에 리뷰 통계 검증까지 추가

dye0p committed:
refactor: 멘토링 카드 조회(findMentoringSummaries) 로직 리팩터링

- 메서드 분리
- 추상화 수준 맞추기
- 메서드 호출되는 순서대로 배치
- ReviewStats 기본 생성 정적팩터리 메서드 추가

dye0p committed:
feat: 멘토링 카드 조회(findMentoringSummaries) 메서드 @ Transactional(readOnly = true) 추가

- 메서드 내에서 여러번의 조회가 일어나고 있으므로 데이터의 정합성을 위해서 읽기 전용 트랜잭션 어노테이션 추가

dye0p commented:
@coderabbitai 리뷰해

dye0p commented on backend/fittoring/src/main/java/fittoring/mentoring/business/repository/MentoringRepository.java:
```diff
@@ -12,7 +13,14 @@
+              JOIN res.mentoring m
+             GROUP BY m.id
+            """)
+    List<ReviewStats> findReviewStats();
```
- 리뷰 정보를 가져오기 위해 멘토링, 예약을 모두 조회해서 리뷰를 가져오는 것이 비효율적이라 판단
 - 멘토링 목록 조회 시 발생할 수 있는 N+1 문제와 방지
 
 이러한 이유로 위와같은 JPQL을 작성하였습니다.
 
 `ReviewStats`는 리뷰 통계 정보를 담은 dto 입니다.  DTO Projection을 사용해서 원하는 값을 dto에 매핑하여 반환할 수 있도록 하였습니다.

kimdevv replied:
p3: 복잡해질 수 있는 로직을 JPQL으로 간단하게 조회하도록 설계하신 것이 좋은 것 같습니다👍  그런데 말씀해주신 `멘토링 목록 조회 시 발생할 수 있는 N+1 문제와 방지`는 JPQL을 통해서 어떻게 방지할 수 있는 건지 잘 이해가 되지 않아서 설명 한 번만 부탁드립니다!!

dye0p replied:
리뷰 정보를 조회하려면 예약에서 조회해야 합니다. 
예약을 조회하기 위해서는 멘토링이 필요한데, 이를 위해서는 멘토링을 순회해야 합니다. 
멘토링을 순회하는 과정에서 예약을 조회하면서 리뷰까지 조회하므로 N개의 쿼리가 나가게 됩니다.

하지만 지금과 같이 JPQL에서 JOIN을 사용하여 조회한다면 1개의 쿼리로 멘토링에 대한 리뷰를 조회할 수 있습니다.

dye0p commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -159,23 +161,34 @@ private List<String> getCategoryTitlesByMentoringId(Long mentoringId) {
             String categoryTitle3
     ) {
         List<Mentoring> mentorings = findMentorings(categoryTitle1, categoryTitle2, categoryTitle3);
+        List<ReviewStats> reviewStats = mentoringRepository.findReviewStats();
+        Map<Long, ReviewStats> reviewStatsMap = createReviewStatsMap(reviewStats);
```
리스트로 받은 리뷰 정보를 다시 map으로 변환한 이유는 각 멘토링에 해당하는 리뷰 정보를 매번 순회하지 않고, mentoringId(key)로 빠르게 찾을 수 있게 하기 위해서 map으로 변환 하였습니다.

jye326 replied:
좋은데요

dye0p commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringService.java:
```diff
@@ -213,10 +231,23 @@ private void validateExistCategoryTitle(String categoryTitle) {
+                mentoring.getId());
+    }
+
+    private ReviewStats getReviewStats(Mentoring mentoring, Map<Long, ReviewStats> reviewStatsMap) {
+        return reviewStatsMap.getOrDefault(
+                mentoring.getId(),
+                ReviewStats.defaultOf(mentoring.getId())
+        );
```
멘토링에 해당하는 리뷰는 있을수도 있고, 없을수도 있습니다. 
만약 있는 경우 해당하는 리뷰 정보를 반환하고, 없는 경우 기본 리뷰 정보를 생성합니다 (평점:0.0, 개수:0)

jye326 replied:
좋은데요

dye0p commented:
@coderabbitai 코드에 직접 코멘트를 달아

dye0p committed:
fix: 필드 타입과 상수 타입 불일치 수정

kimdevv reviewed (commented):
고생하셨습니다!! 궁금한 거 하나 커멘트로 남깁니다! 코드래빗의 리뷰도 좋은 것 같네요~

ppparkta reviewed (commented):
고생하셨습니다. 지금 작성해주신 방식 좋은 것 같아요. 그런데 제가 작성했던 부분과 상당부분 겹치는 것 같아요 논의해서 한쪽으로 통일합시다. 오늘 이야기해봐요!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/dto/ReviewStats.java:
```diff
@@ -0,0 +1,15 @@
+package fittoring.mentoring.business.service.dto;
+
+public record ReviewStats(
+        Long mentoringId,
+        Double reviewAverage,
+        long reviewCount
+) {
```
p1: 이 부분 변수명을 회의시간에 의견 나눈대로 ratingAverage와 ratingCount로 변경 부탁드립니다. 

더불어 레코드명에 Dto접미사를 붙여주시면 좋을 것 같습니다!

dye0p replied:
확인했습니다! 반영하겠습니다.

ppparkta reviewed (approved):
리뷰남긴 부분 잘 수정해주실 것 같아서 승인합니다. 오전에 코드설명회를 통해 충분히 이해되었습니다. 설명해주셔서 감사합니다!

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/dto/MentoringSummaryResponse.java:
```diff
@@ -11,12 +12,15 @@ public record MentoringSummaryResponse(
         String profileImageUrl,
-        String introduction
+        String introduction,
+        double reviewAverage,
```
p2: `reviewAverage` 보다는 `averageRating` 은 어떨까요?

ppparkta replied:
일관성 위해서 ratingAverage로 부탁드립니다~~

dye0p replied:
rating- 으로 일괄 통일하였습니다!

dye0p committed:
Merge branch 'develop' into feature/477-fix-mentoring-summary-response

# Conflicts:
#	backend/fittoring/src/test/java/fittoring/integration/mentoring/api/MentoringControllerTest.java

dye0p committed:
refactor: 리뷰 통계 리스트 조회 쿼리메서드 ReviewRepository로 이동

dye0p committed:
refactor: 리뷰통계 리스트 쿼리 메서드의 매개변수 인자로 mentoringIds를 받도록 추가

- 필요한 mentoringId로만 리뷰 통계를 찾을 수 있도록 하기 위함

dye0p committed:
refactor: ReviewStats 객체를 제거하고, RatingStatsDto를 리뷰 통계 dto로 사용할 수 있도록 통일

dye0p committed:
fix: findReviewStatsByMentoringIds 쿼리 메서드의 반환 dto 수정

dye0p committed:
refactor: findRatingStatsByMentoringId 쿼리메서드 반환 옵셔널 처리

- 단건 조회이므로 옵셔널로 처리하여 null에 대한 처리를 해줄 수 있다.

dye0p committed:
refactor: findRatingStatsByMentoringId 쿼리메서드 반환 옵셔널 처리

- 단건 조회이므로 옵셔널로 처리하여 null에 대한 처리를 해줄 수 있다.

dye0p committed:
test: 리뷰 통계 사용 테스트에서 RatingStatsDto를 사용하도록 수정

dye0p committed:
refactor: 멘토링 상세 조회 MentoringResponse 생성 로직 수정

- MentoringResponse 내부에서 image에 대한 null 검사를 해주고 있으므로 service 외부에서 하던 image null 검사를 제거하고 하나의 정적 팩터리 메서드로 생성하도록 수정

dye0p committed:
refactor: 이미지 find 로직 메서드 분리 및 호출 순으로 배치

dye0p committed:
refactor: 멘토링 카드 정보 응답 리뷰 통계 필드 네이밍 변경

dye0p committed:
fix: 멘토링 카드 조회 응답 리뷰 평점 필드 타입 변경(double -> String)

- 소수점 1자리까지 제한하기 위해서 String 타입으로 포맷팅하기 위함

dye0p committed:
Merge branch 'develop' into feature/477-fix-mentoring-summary-response

# Conflicts:
#	backend/fittoring/src/test/java/fittoring/integration/mentoring/api/MentoringControllerTest.java
